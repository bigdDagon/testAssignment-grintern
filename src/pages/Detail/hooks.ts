import { useQuery } from "@tanstack/react-query";

import { fetchBook } from "./services";

export const useFetchBook = (id: string | undefined) => {
  return useQuery(
    [`book-${id}`],
    async () => {
      const { data } = await fetchBook(id);
      return data.results;
    },
    { staleTime: 60000, refetchOnWindowFocus: false }
  );
};
