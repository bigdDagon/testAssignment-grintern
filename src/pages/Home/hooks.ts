import { useQuery } from "@tanstack/react-query";

import { fetchBookLists, searchBook } from "./services";
import { Book } from "../../types/types";

export const useFetchBookLists = () => {
  return useQuery(
    ["bookLists"],
    async () => {
      const {
        data: { results },
      } = await fetchBookLists();
      return results;
    },
    { refetchOnWindowFocus: false, staleTime: Infinity }
  );
};

export const useSearchBook = (searchTerm: string) => {
  const encoded = encodeURIComponent(searchTerm);
  return useQuery(
    [searchTerm],
    async () => {
      const { data } = await searchBook(encoded);
      return data;
    },
    {
      enabled: !!searchTerm,
      select: (data) => {
        const results = data.results.map((book: Book) => ({
          value: book.id,
          option: book.title,
        }));
        return { results, count: data.count };
      },
    }
  );
};
