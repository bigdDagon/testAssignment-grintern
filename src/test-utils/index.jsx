import { render } from "@testing-library/react";
import { Router } from "react-router-dom";

import { ThemeProvider } from "@mui/material/styles";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { ReservationContext } from "../context/reservationContext";
import { theme } from "../theme";
import { defaultReservationProps } from "./fixtures";

const Harness =
  ({ reservationContext }) =>
  ({ children }) => {
    const queryClient = new QueryClient();
    return (
      <QueryClientProvider client={queryClient}>
        <ReservationContext.Provider
          value={{ reservedList: reservationContext }}
        >
          <ThemeProvider theme={theme}>
            <Router location={"/"}>{children}</Router>
          </ThemeProvider>
        </ReservationContext.Provider>
      </QueryClientProvider>
    );
  };

const customRender = (
  ui,
  { reservationContext = defaultReservationProps } = {},
  options
) => {
  const Wrapper = Harness({ reservationContext });
  return {
    ...render(ui, { wrapper: Wrapper, ...options }),
  };
};

export * from "@testing-library/react";

export { customRender as render };
