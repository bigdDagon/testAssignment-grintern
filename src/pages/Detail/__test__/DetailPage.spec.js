import { useQuery } from "@tanstack/react-query";

import DetailPage from "..";
import { mockBookDetail } from "../../../test-utils/fixtures";
import { render, screen, fireEvent } from "../../../test-utils";
import { useFetchBook } from "../hooks";

jest.mock("../hooks");

const mockedUseFetchBook = useFetchBook;

describe("Detail Page", () => {
  it("displays loader", () => {
    mockedUseFetchBook.mockImplementation(() => ({
      data: [],
      isFetching: true,
    }));
    render(<DetailPage />);
    expect(screen.getByTestId("loader")).toBeVisible();
  });

  it("displays error message", () => {
    mockedUseFetchBook.mockImplementation(() => ({
      data: [],
      isFetching: false,
      isError: true,
    }));
    render(<DetailPage />);
    expect(screen.getByTestId("error-page")).toBeVisible();
  });

  it("renders Detail Page", async () => {
    mockedUseFetchBook.mockImplementation(() => ({
      data: mockBookDetail,
      isFetching: false,
      isError: false,
    }));

    render(<DetailPage />);

    const detailPage = screen.getByTestId("detail-page-wrapper");
    expect(detailPage).toBeVisible();
    expect(screen.getByRole("button", { name: "Back" })).toBeVisible();

    expect(screen.getByTestId("cover-image")).toBeVisible();
    expect(screen.getByTestId("Author")).toBeVisible();
    expect(screen.getByTestId("Subject")).toBeVisible();
    expect(screen.getByTestId("Summary")).toBeVisible();

    expect(screen.getByTestId("book-title")).toBeVisible();

    const reserveButton = screen.getByRole("button", { name: "Reserve" });

    expect(reserveButton).toBeInTheDocument();

    expect(detailPage).toMatchSnapshot();
  });
});
