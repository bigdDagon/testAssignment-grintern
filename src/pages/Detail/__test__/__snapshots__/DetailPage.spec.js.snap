// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Detail Page renders Detail Page 1`] = `
<div
  class="MuiBox-root css-79elbk"
  data-testid="detail-page-wrapper"
>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-11jvbna-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    Back
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
  <div
    class="MuiBox-root css-kwuok"
  >
    <div
      class="MuiBox-root css-opcd4z"
      data-testid="cover-image"
    >
      <img
        alt="Romeo and Juliet"
        src="https://www.gutenberg.org/cache/epub/1513/pg1513.cover.medium.jpg"
        width="100%"
      />
    </div>
    <div
      class="MuiBox-root css-i9gxme"
    >
      <div
        class="MuiBox-root css-0"
      >
        <h4
          class="MuiTypography-root MuiTypography-h4 css-1blr37w-MuiTypography-root"
          data-testid="book-title"
        >
          Romeo and Juliet
        </h4>
        <div
          class="MuiBox-root css-178yklu"
        >
          <div
            class="MuiBox-root css-68zlnx"
            data-testid="Author"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-5uz6um-MuiTypography-root"
            >
              Author
              :
            </p>
            <div
              class="MuiBox-root css-d0uhtl"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-5uz6um-MuiTypography-root"
              >
                Shakespeare, William (1564 ~ 1616)
              </p>
            </div>
          </div>
          <div
            class="MuiBox-root css-68zlnx"
            data-testid="Subject"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-5uz6um-MuiTypography-root"
            >
              Subject
              :
            </p>
            <div
              class="MuiBox-root css-d0uhtl"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-5uz6um-MuiTypography-root"
              >
                Conflict of generations -- Drama
              </p>
              <p
                class="MuiTypography-root MuiTypography-body1 css-5uz6um-MuiTypography-root"
              >
                Juliet (Fictitious character) -- Drama
              </p>
              <p
                class="MuiTypography-root MuiTypography-body1 css-5uz6um-MuiTypography-root"
              >
                Romeo (Fictitious character) -- Drama
              </p>
              <p
                class="MuiTypography-root MuiTypography-body1 css-5uz6um-MuiTypography-root"
              >
                Tragedies
              </p>
              <p
                class="MuiTypography-root MuiTypography-body1 css-5uz6um-MuiTypography-root"
              >
                Vendetta -- Drama
              </p>
              <p
                class="MuiTypography-root MuiTypography-body1 css-5uz6um-MuiTypography-root"
              >
                Verona (Italy) -- Drama
              </p>
              <p
                class="MuiTypography-root MuiTypography-body1 css-5uz6um-MuiTypography-root"
              >
                Youth -- Drama
              </p>
            </div>
          </div>
          <div
            class="MuiBox-root css-68zlnx"
            data-testid="Summary"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-5uz6um-MuiTypography-root"
            >
              Summary
              :
            </p>
            <div
              class="MuiBox-root css-d0uhtl"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-5uz6um-MuiTypography-root"
              >
                Lorem Ipsum Lorem Ipsum Lorem IpsumLorem IpsumLorem IpsumLorem IpsumLorem Ipsum Lorem Ipsum Lorem   IpsumLorem Ipsum vLorem IpsumLorem Ipsum Lorem Ipsum Lorem Ipsum Lorem IpsumLorem IpsumLorem Ipsum   Lorem IpsumLorem Ipsum Lorem Ipsum Lorem IpsumLorem Ipsum vLorem IpsumLorem Ipsum
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="MuiBox-root css-1d3apxl"
  >
    <button
      class="MuiButtonBase-root MuiFab-root MuiFab-extended MuiFab-sizeLarge MuiFab-secondary MuiFab-root MuiFab-extended MuiFab-sizeLarge MuiFab-secondary css-1uhx2if-MuiButtonBase-root-MuiFab-root"
      tabindex="0"
      type="button"
    >
      Reserve
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
